File: 0525-contiguous-array/0525-contiguous-array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        int n=nums.size(),sum=n,ans=0,shift=n;
        vector<int>mp(2*n+3,-2);
        mp[sum]=-1;
        for(int i=0;i<n;i++)
        {
            sum+=nums[i]*2-1;
            if(mp[sum]!=-2)
                ans=max(ans,i-mp[sum]);
            else
                mp[sum]=i;
        }
        return ans;
    }
};

File: 0525-contiguous-array/0525-contiguous-array.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        prefIndex = defaultdict(int)
        prefIndex[0] = -1
        pref = ans = 0
        for i in range(len(nums)):
            pref += nums[i] * 2 - 1
            if pref in prefIndex:
                ans = max(ans, i - prefIndex.get(pref, i))
            else:
                prefIndex[pref] = i
        return ans
